### DISCUSSION TITLE
group by edge case in psycopg (works under psycopg2)

### DISCUSSION THREAD

**Discussion by guruofgentoo** [sqlalchemy/sqlalchemy#11854] [Usage Questions] [answered]
As reported in a [psycopg bug report ](https://github.com/psycopg/psycopg/issues/874), if:
- a column is used in an expression for a query containing aggregates
- and, the same expression is used for grouping

Then, the resulting execution raises an exception for the column missing from grouping, even though it is present in the expression.

The example code in the linked issue sufficiently matches my own case. The psycopg maintainer closed the issue with an indication that this is a sqlalchemy problem, but I don't see an issue or discussion related to it here.

**Comment by CaselIT** [comment]
Hi,

Either use the same expression or use a literal string in the groupby to refer to a particular column in the select list.

By same expression I mean
```py
dt = func.date_trunc("week", User.created_at)
stmt = select(
    dt.label("date"),
    func.count(User.id).label("count")
).group_by(dt)
```

this also works
```py
stmt = select(
    func.date_trunc("week", User.created_at).label("date"),
    func.count(User.id).label("count")
).group_by("date")
```

**Comment by guruofgentoo** [comment]
Using a label and a matching literal string for grouping is the proposed workaround in the linked issue. And that does work, for avoiding the exception.

However, doing so seems messy compared to being able to just use an expression without explicitly labeling it. For composed queries of greater complexity, this means keeping track of both labels and expressions, vs just the expressions.

**Reply by CaselIT** [reply]
You can keep track of just the expression. See the first options above

**Comment by CaselIT** [comment]
as to why postgresql complains when you do this
```py
stmt = select(
    func.date_trunc("week", User.created_at)
).group_by(func.date_trunc("week", User.created_at))
```
is because sqlalchemy uses bound parameters unless told not to, so the above is equivalent to
```py
stmt = select(
    func.date_trunc(literal("week"), User.created_at)
).group_by(func.date_trunc(literal("week"), User.created_at))
```

The two literal there are not related in any way, so two bound parameters are passed to postgresql. This makes postgresql complain since it cannot verify that the two bound parameters always have the same value.

To tell sqlalchemy to use the same bound parameter for a literal share the same instance.
This will make use of the same bound parameter for the two functions, making postgresql happy since it see that the same value is used both in the select list and in the groubpy list.
```py
week = literal("week")
stmt = select(
    func.date_trunc(week , User.created_at)
).group_by(func.date_trunc(week, User.created_at))
```

**Comment by guruofgentoo** [comment]
I see what's happening underneath now, so documenting for posterity. Apparently, psycopg2 interpolated parameters, but the more recent psycopg does not. Because the parameterized query results in a difference in the resulting sql, the expressions are seen as distinct, and therefore invalid for GROUP BY matching.

For example:

```
from sqlalchemy import create_engine, select, sql, Column, Date, Integer, Unicode
from sqlalchemy.ext.hybrid import hybrid_property
from sqlalchemy.orm import sessionmaker
from sqlalchemy.orm import declarative_base
from sqlalchemy_utils import ArrowType

class Base:
    pass

Base = declarative_base(cls=Base)


class WorkSchedule(Base):
    __tablename__ = 'work_schedules2'

    id = Column(Integer, primary_key=True)

    work_date = Column(Date, nullable=False)
    start_time = Column(ArrowType)

    user_name = Column(Unicode, nullable=False)

    @hybrid_property
    def week_base_date(self):
        """Get the base date for a work week, so we can group dates by week."""
        return arrow.get(self.work_date).shift(days=-self.work_date.weekday()).date()

    @week_base_date.expression
    def week_base_date(cls):
        dow = sql.cast(sql.extract('dow', cls.work_date), Integer)
        return cls.work_date - ((6 + dow) % 7)

    @classmethod
    def query_worked_days(cls):
        return select(
            cls.user_name,
            cls.week_base_date,
            sql.functions.sum(sql.case((cls.start_time.is_(None), 0), else_=1)).label(
                'worked_days'
            ),
        ).group_by(cls.user_name, cls.week_base_date)


engine_psycopg = create_engine("postgresql+psycopg://postgres@localhost:12432/iriba-maps")
engine_psycopg2 = create_engine("postgresql://postgres@localhost:12432/iriba-maps")
Session = sessionmaker(bind=engine_psycopg)
Base.metadata.create_all(engine_psycopg)
session = Session()

session.execute(WorkSchedule.query_worked_days()).all()
```

We get a few different SQL outputs from these:
```
psycopg2:
SELECT work_schedules2.user_name AS work_schedules2_user_name, work_schedules2.work_date - (6 + CAST(EXTRACT(dow FROM work_schedules2.work_date) AS INTEGER)) % 7 AS week_base_date, sum(CASE WHEN (work_schedules2.start_time IS NULL) THEN 0 ELSE 1 END) AS worked_days 
FROM work_schedules2 GROUP BY work_schedules2.user_name, work_schedules2.work_date - (6 + CAST(EXTRACT(dow FROM work_schedules2.work_date) AS INTEGER)) % 7

psycopg without labeling:
SELECT work_schedules2.user_name AS work_schedules2_user_name, work_schedules2.work_date - ($1::INTEGER + CAST(EXTRACT(dow FROM work_schedules2.work_date) AS INTEGER)) % $2::INTEGER AS week_base_date, sum(CASE WHEN (work_schedules2.start_time IS NULL) THEN $3::INTEGER ELSE $4::INTEGER END) AS worked_days 
FROM work_schedules2 GROUP BY work_schedules2.user_name, work_schedules2.work_date - ($5::INTEGER + CAST(EXTRACT(dow FROM work_schedules2.work_date) AS INTEGER)) % $6::INTEGER

psycopg with labeling:
SELECT work_schedules2.user_name, work_schedules2.work_date - ($1::INTEGER + CAST(EXTRACT(dow FROM work_schedules2.work_date) AS INTEGER)) % $2::INTEGER AS week_base_date, sum(CASE WHEN (work_schedules2.start_time IS NULL) THEN $3::INTEGER ELSE $4::INTEGER END) AS worked_days 
FROM work_schedules2 GROUP BY work_schedules2.user_name, week_base_date
DETAIL:  parameters: $1 = '6', $2 = '7', $3 = '0', $4 = '1'
```

From a modeling/querying perspective, the power of `hybrid_property` is that I can use it anywhere in the same way as an actual column, thereby composing queries without having to always remember the underlying pieces. This power is subverted when the parameter matching fails to be exact, i.e. requiring the developer to know which expressions can be used plainly, and which require labeling to proceed. Not saying either way is right or wrong, just that the path was smoother under psycopg2.

Given that named parameters are supported, would it be possible to use a set of named parameters instead of positional, with any usage of a hybrid property resulting in the same parameter set?

**Reply by CaselIT** [reply, marked as answer]
See my other reply.
Alternatively use client side cursors with psycopg3 https://www.psycopg.org/psycopg3/docs/advanced/cursors.html#client-side-binding-cursors -- but I don't know if there are caveat here.

**Reply by guruofgentoo** [reply]
Thanks - using ClientCursor was the missing piece I was looking for.

**Reply by CaselIT** [reply]
I would suggest using the other approach clarified here https://github.com/sqlalchemy/sqlalchemy/discussions/11854#discussioncomment-10593344, but whatever works I guess.
